plugins/google/pkg/googleclient/google.go

```diff
@@ -0,0 +1,178 @@
+package googleclient
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	admin "google.golang.org/api/admin/directory/v1"
+	"google.golang.org/api/option"
+)
+
+type GoogleClient struct {
+	googleClient *admin.Service
+}
+
+func GetRefreshToken(ctx context.Context, clientID, clientSecret string, port int) (string, error) {
+	redirectURL := fmt.Sprintf("http://localhost:%d", port)
+	// Configure the OAuth 2.0 client credentials
+	config := &oauth2.Config{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		RedirectURL:  redirectURL,
+		Scopes: []string{
+			admin.AdminDirectoryUserScope,
+			admin.AdminDirectoryGroupScope,
+		},
+		Endpoint: google.Endpoint,
+	}
+
+	// Create an HTTP server for handling the OAuth 2.0 callback
+	server := &http.Server{Addr: fmt.Sprintf(":%d", port), ReadHeaderTimeout: 5 * time.Second}
+
+	// Create a channel to receive the authorization code
+	authCodeChan := make(chan string)
+
+	// Start the HTTP server in a separate goroutine
+	go func() {
+		http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+			code := r.URL.Query().Get("code")
+			fmt.Fprintf(w, "Authorization code received. You can close this tab now.")
+			authCodeChan <- code
+		})
+
+		if err := server.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			log.Fatalf("Failed to start HTTP server: %v", err)
+		}
+	}()
+
+	// Generate the authorization URL with "access_type=offline" and "prompt=consent"
+	authURL := config.AuthCodeURL("state-token", oauth2.AccessTypeOffline, oauth2.SetAuthURLParam("prompt", "consent"))
+
+	fmt.Printf("Go to the following URL to authorize the application:\n\n%s\n\n", authURL)
+	fmt.Println("Waiting for authorization code...")
+
+	// Receive the authorization code from the channel
+	authCode := <-authCodeChan
```

Comment:
If I understand correctly, the http server is only intended to stay up until the redirect completes and we have a token.
If so, you should call `server.Shutdown(ctx)` to stop it gracefully (or `server.Close()` to kill it) before exiting this function.

Reply:
this is likely due to my own ignorance of how go routines and channels work, but when I tried to but a `server.Shutdown(ctx)` (or even a `defer` of that) in the caller, after the `authCode` was obtained from the channel, I seem to get a deadlock - browser shows that a code has been sent but it never gets printed out (i.e. the program hangs).

Reply:
I don't think it's related to the shutdown deadlock you hit but I think you want to call `close(authCodeChan)` here after reading the auth code.

Did you put the call to `server.Shutdown` in the body of `GetRefreshToken` just below this line?
I'd try to see of calling `server.Close()` also causes the program to hang.